networking/notes/shelley/BlockRetrievalRecovery.md

```diff
@@ -0,0 +1,126 @@
+# Block retrieval and recovery mode over diffusion layer
+
+This document shows how the interface proposed [here](./Interface.md) can
+be used to implement the current block retrieval and recovery mode system
+without significant changes. We will eventually make significant changes, but
+it would be nice to shim the diffusion layer interface in as a separate work
+item before making other improvements.
+
+Let's go in order of a block header's journey through the system. It starts
+when it enters the block retrieval queue.
+
+## `handleBlockHeaders`
+
+Block headers enter the system here. It's a time-warp listener.
+`handleBlockHeaders` calls `handleUnsolicitedHeaders` (which in
+fact only works if there's precisely one header), classifying the
+header (continues, alternative, useless, or invalid) according to current
+state (our tip) and dropping it into the block retrieval queue if
+it's not useless or invalid.
+
+This is fine, all we're doing is changing the source of the block headers.
+Instead of coming from an explicit time-warp conversation, they'll come
+from the diffusion layer (which uses a time-warp conversation, but that's
+abstracted). We can get rid of `handleUnsolicitedHeaders` because all
+we need is `handleUnsolicitedHeader`, which can be called whenever the
+diffusion layer presents a block header. `handleUnsolicitedHeader` requires
+a `NodeId` of the provenance of the block, but that's only there so it
+knows who to ask for the body, so it should be removed because it's the
+diffusion layer's responsibility to know where to find the body.
```

Comment:
@dcoutts said

> And are we confident that the block cache layer can indeed do this. Will it need to track where it knows it can fetch new blocks from?

@avieth said

> Yes I imagine that's how it would work. Included in the cache would be a set of NodeIds for peers which have announced that header.